avr/cores/megacommand/MCL/MDSeqTrackData.h

```diff
@@ -22,31 +28,69 @@ class MDSeqTrackData_270 {
   uint8_t length;
   uint8_t speed;
   uint32_t slide_mask32; // to be increased to 64bits
-  uint8_t locks[NUM_MD_LOCKS][NUM_MD_STEPS];
-  uint8_t locks_params[NUM_MD_LOCKS];
+  uint8_t locks[NUM_MD_LOCKS_270][NUM_MD_STEPS_270];
+  uint8_t locks_params[NUM_MD_LOCKS_270];
   uint64_t pattern_mask;
   uint64_t lock_mask;
-  uint8_t conditional[NUM_MD_STEPS];
-  uint8_t timing[NUM_MD_STEPS];
+  uint8_t conditional[NUM_MD_STEPS_270];
+  uint8_t timing[NUM_MD_STEPS_270];
+};
+
+class MDSeqStepDescriptor {
+public:
+  uint8_t locks; // <-- bitfield of 8 locks in the current step, first lock is lsb
+  bool locks_enabled : 1; // true: locks are enabled. false: locks are disabled, but still occupy lock slots
+  bool trig : 1;
+  bool slide : 1;
+  bool cond_plock : 1;
+  uint8_t cond_id : 4;
+  bool is_lock_bit(const uint8_t idx) const { return locks & (1 << idx); }
+  bool is_lock(const uint8_t idx) const { return is_lock_bit(idx) && locks_enabled; }
 };
 
 class MDSeqTrackData {
 public:
-  uint32_t slide_mask32; // to be increased to 64bits
-  uint8_t locks[NUM_MD_LOCKS][NUM_MD_STEPS];
+  uint8_t locks[NUM_MD_LOCK_SLOTS];
   uint8_t locks_params[NUM_MD_LOCKS];
-  uint64_t pattern_mask;
-  uint64_t lock_mask;
-  uint8_t conditional[NUM_MD_STEPS];
   uint8_t timing[NUM_MD_STEPS];
-  void init() {
-    memset(&locks, 0, NUM_MD_LOCKS * NUM_MD_STEPS);
-    memset(&locks_params, 0, NUM_MD_LOCKS);
-    pattern_mask = 0;
-    lock_mask = 0;
-    memset(&conditional, 0, NUM_MD_STEPS);
-    memset(&timing, 0, NUM_MD_STEPS);
+  MDSeqStepDescriptor steps[NUM_MD_STEPS];
+
+  // get the pointer to the data chunk.
+  // useful to skip the vtable
+  void* data() const { return locks; }
```

Comment:
and this ? should the return be a pointer ?

Reply:
ah, yes... wondering how it gets through the type check.

Reply:
here locks _IS_ a pointer.

Reply:
``` 
 51 class MDSeqTrackData {
 52 public:
 53   uint8_t locks[NUM_MD_LOCK_SLOTS];
```
?

Reply:
ah because it's an array.